pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,256 @@
+			BatchingMaxPublishDelay: time.Millisecond * 10,
+			// set send timeout to be infinity to prevent potential deadlock with consumer
+			// that might happen when consumer is blocked due to unacked messages
+			SendTimeout: 0,
+		})
+		if err != nil {
+			log.Errorf("create producer error:%s", err.Error())
+		}
+		return err
+	}
+	return nil
+}
+
+func (gi *GoInstance) setupConsumer() chan pulsar.ConsumerMessage {
+	subscriptionType := pulsar.Shared
+	if int32(gi.context.InstanceConf.FuncDetails.Source.SubscriptionType) == pb.SubscriptionType_value["FAILOVER"] {
+		subscriptionType = pulsar.Failover
+	}
+
+	funcDetails := gi.context.InstanceConf.FuncDetails
+	subscriptionName := funcDetails.Tenant + "/" + funcDetails.Namespace + "/" + funcDetails.Name
+
+	properties := getProperties(getDefaultSubscriptionName(
+		funcDetails.Tenant,
+		funcDetails.Namespace,
+		funcDetails.Name), gi.context.InstanceConf.InstanceID)
+
+	channel := make(chan pulsar.ConsumerMessage)
+
+	var (
+		consumer pulsar.Consumer
+		err      error
+	)
+
+	for topic, consumerConf := range funcDetails.Source.InputSpecs {
+		log.Debugf("Setting up consumer for topic: %s with subscription name: %s", topic, subscriptionName)
+		if consumerConf.ReceiverQueueSize != nil {
+			if consumerConf.IsRegexPattern {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:     topic,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              subscriptionType,
+					MessageChannel:    channel,
+				})
+			} else {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:             topic,
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              subscriptionType,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					MessageChannel:    channel,
+				})
+			}
+		} else {
+			if consumerConf.IsRegexPattern {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:    topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             subscriptionType,
+					MessageChannel:   channel,
+				})
+			} else {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:            topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             subscriptionType,
+					MessageChannel:   channel,
+				})
+
+			}
+		}
+
+		if err != nil {
+			log.Errorf("create consumer error:%s", err.Error())
+		}
+		gi.consumers[topic] = consumer
+	}
+	return channel
+}
+
+func (gi *GoInstance) handlerMsg(input pulsar.Message) (output []byte, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = NewContext(ctx, gi.context)
+	msgInput := input.Payload()
+	output, err = gi.function.Process(ctx, msgInput)
```

Comment:
`return gi.function.Process(ctx, msgInput)` and let the caller handle the error.